### ISSUE TITLE
Edit Starters doesn't properly handle removal of all dependencies

### ISSUE THREAD

**Issue by kdvolder** [spring-projects/sts4#52]
When all dependencies are removed/unchecked from a boot project. The edit starters tends to create a broken project because it does not ensure that at leats a dependency like:

```
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter</artifactId>
</dependency>
```

Steps to reproduce:

1. create a project with "new Spring starter". Select a single dependency (e.g. `web`).
2. Use "Edit starters" and remove the dependency.

=> The dependency is removed from the pom leaving no dependencies in the pom at all. 

The project is now broken and has compile errors.

Expected behavior is that upon removal of the last dependency we ensure that at least a dependency on 
the 'base' `spring-boot-starter` dependency exists. If it doesn't it should be added.

**Comment by Eskibear** [comment]
The expected behavior is not enough, a corner case is found. 

 The starter `devtools` only introduces following dependency in **runtime** scope:
```
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-devtools</artifactId>
      <scope>runtime</scope>
    </dependency>
```
So we still need to add the base starter back if it's the only starter of the project.

Ref:
https://github.com/Microsoft/vscode-spring-initializr/issues/64

**Comment by kdvolder** [comment]
@Eskibear 

I just got a report from another user in slack chat, complaining about other discrepancies between edit starters and what intializer app generates.

Apparantly since this was implemented in STS quite a while ago, the initializer app has gotten a bit more sophisticated and it is not just a 1-1 mapping between a selected checkbox and a dependency. There are also some dependencies that get added as a kind of 'cross selection'. I.e. sometimes extra dependencies will get added as 'extra' when two boxes are selected together.

The example he gave was: 

> If I select `RabbitMQ` and `Cloud Stream` on a new project, I get the amqp starter, `spring-cloud-stream`, `spring-cloud-stream-binder-rabbit` and `spring-cloud-stream-test-support`.

If you instead one only selects `RabbitMQ` intially. Then add 'Cloud Stream' via edit starters then  `spring-cloud-stream-binder-rabbit` and `spring-cloud-stream-test-support` are not being added.

I'm not totally sure yet how we can handle this. But admittedly I think it should be thought of as a bug.

I have some ideas that perhaps the 'Edit Starters' should be re-written from scratch in a way that it doesn't try to decide by itself how to create the edits. Instead, I maybe we could use initializr app to create two poms. One for the 'intiali' state of the edit starters dialog selection, and then a second one for the edited state. Then compute some kind of a diff between the two and use that diff to update the user's pom. I think this might be a better way as it wouldn't need to know how the initializer logic works internally.

**Comment by Eskibear** [comment]
Yes we also noticed the similar scenarios. I think the best way is to have related APIs telling the exact mapping of dependencies. Diffing XML content is also acceptable. But plain textual diff also has some pitalls, e.g. compare the following two BOMs :

```
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.microsoft.azure</groupId>
        <artifactId>azure-spring-boot-bom</artifactId>
        <version>2.0.1</version>      <!-- HERE -->
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```


```
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <azure.version>2.0.1</azure.version>       <!-- HERE -->
        <java.version>1.8</java.version>
    </properties>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.microsoft.azure</groupId>
                <artifactId>azure-spring-boot-bom</artifactId>
                <version>${azure.version}</version>      <!-- HERE -->
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
```

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Martin Lippert:) 

I accepted this story, seems to work fine, but uncovered a little bug in the edit starters wizard, please also take a look at bug item #158427358. Thanks!!!
